### PR TITLE
[WIP] feat: Add source maps

### PR DESCRIPTION
**Pull request by Marwes**
Resurrected another old branch which adds source maps. Currently this always creates the map which may not be desirable but it should be very low overhead and optional creation may be added later. Going to experiment with this a bit in the language server so I may need to tweak the interface of this before we can merge this.

### CODE REVIEW

File: vm/src/compiler.rs [hunk 1]
```diff
@@ -312,14 +322,16 @@ impl<'a, T: CompilerEnv> CompilerEnv for &'a T {
 impl<'a> Compiler<'a> {
     pub fn new(globals: &'a (CompilerEnv + 'a),
                vm: &'a GlobalVmState,
-               symbols: SymbolModule<'a>)
+               symbols: SymbolModule<'a>,
+               source_map: &'a Source<'a>)
```
Comment (brendanzab) [inline vm/src/compiler.rs]: This isn't really a `SourceMap` at this stage - right?

File: src/compiler_pipeline.rs [hunk 2]
```diff
@@ -196,6 +199,7 @@ impl<E, Extra> Compileable<Extra> for TypecheckValue<E>
                compiler: &mut Compiler,
                thread: &Thread,
                filename: &str,
+               expr_str: &str,
```
Comment (brendanzab) [inline src/compiler_pipeline.rs]: Wondering if we could be passing in a `Source` at this stage. Ditto with the other methods.

File: src/compiler_pipeline.rs [hunk 3]
```diff
@@ -181,10 +183,11 @@ impl<'a, 'b, T> Compileable<(&'a str, Option<&'b ArcType>)> for T
                compiler: &mut Compiler,
                thread: &Thread,
                file: &str,
-               (expr_str, expected_type): (&'a str, Option<&'b ArcType>))
+               expr_str: &str,
```
Comment (brendanzab) [inline src/compiler_pipeline.rs]: :+1:

File: src/compiler_pipeline.rs [hunk 4]
```diff
@@ -181,10 +183,11 @@ impl<'a, 'b, T> Compileable<(&'a str, Option<&'b ArcType>)> for T
...
+               expected_type: Option<&'b ArcType>)
                -> Result<CompileValue<Self::Expr>> {
         self.typecheck_expected(compiler, thread, file, expr_str, expected_type)
-            .and_then(|tc_value| tc_value.compile(compiler, thread, file, ()))
+            .and_then(|tc_value| tc_value.compile(compiler, thread, file, expr_str, ()))
     }
 }
 impl<E, Extra> Compileable<Extra> for TypecheckValue<E>
```
Comment (brendanzab) [inline src/compiler_pipeline.rs]: Could ad a space here while you are in this file. Same with the other `impl`s.

File: vm/src/source_map.rs [hunk 5]
```diff
@@ -0,0 +1,37 @@
+use base::pos::Line;
+
+#[derive(Debug)]
+pub struct SourceMap {
+    /// `map` stores the index of the first instruction for each line.
```
Comment (brendanzab) [inline vm/src/source_map.rs]:
```rust
/// The index of the first instruction for each line
```

File: vm/src/compiler.rs (hunk 1 above)
Comment (Marwes) [inline vm/src/compiler.rs]: It is some sort of source mapping however :P I should rename it now however.

File: src/compiler_pipeline.rs (hunk 3 above)
Comment (Marwes) [inline src/compiler_pipeline.rs]: Its still pretty terrible however :/

File: vm/src/compiler.rs (hunk 1 above)
Comment (brendanzab) [inline vm/src/compiler.rs]: Yeah... it's interesting that they are similar in function. Let's see how it pans out - maybe in the future we can find some common idea in `Source` and `SourceMap` to make this stuff simpler.

**Comment by Marwes** [comment]
Superseeded by #182

**Pull request closed without merging**